metadata_writer_nne.c

```diff
@@ -94,132 +309,725 @@ static uint8_t md_nne_handle_gps_event(struct md_writer_nne *mwn,
+        "CONN_EVENT_META_MODE_UPDATE",
+        "CONN_EVENT_META_QUALITY_UPDATE",
+    };
+
+    META_PRINT_SYSLOG(mwn->parent, LOG_ERR, "NNE writer: md_conn_event %s: interface_name=%s, imei=%s, network_address=%s, event_value=%d, event_value_str=%s",
+                      conn_event_type[mce->event_param],  mce->interface_name, mce->imei, mce->network_address, mce->event_value, mce->event_value_str);
+}*/
+
+static struct nne_modem *md_nne_get_modem(struct nne_modem_list *modem_list, uint32_t mccmnc)
+{
+    struct nne_modem *e = NULL;
+    for (e = modem_list->lh_first; e != NULL; e = e->entries.le_next)
+        if (e->mccmnc == mccmnc)
+            return e;
+
+    return NULL;
+}
+
+struct nne_value nne_value_init_str(char* str)
+{
+    struct nne_value value;
+    value.type = NNE_TYPE_STRING;
+    value.u.v_str = str;
+    return value;
+}
+
+struct nne_value nne_value_init(enum nne_type type, void *ptr, int offset)
+{
+    struct nne_value value;
+    value.type = type;
+    switch(type)
+    {
+        case NNE_TYPE_NULL:
+            break;
+        case NNE_TYPE_INT8:
+            value.u.v_int8 = *(int8_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_UINT8:
+            value.u.v_uint8 = *(uint8_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_INT16:
+            value.u.v_int16 = *(int16_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_UINT16:
+            value.u.v_uint16 = *(uint16_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_INT32:
+            value.u.v_int32 = *(int32_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_UINT32:
+            value.u.v_uint32 = *(uint32_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_STRING:
+            value.u.v_str = (char*)ptr + offset;
+            break;
+    }
+
+    return value;
+}
+
+static int nne_value_compare(struct nne_value v1, struct nne_value v2)
+{
+    if (v1.type != v2.type)
+        return -2;
+
+    switch(v1.type)
+    {
+        case NNE_TYPE_NULL:
+            return 0; // NULL equals NULL
+        case NNE_TYPE_INT8:
+            return (v1.u.v_int8 == v2.u.v_int8) ? 0 : ((v1.u.v_int8 < v2.u.v_int8) ? -1 : 1);
+        case NNE_TYPE_UINT8:
+            return (v1.u.v_uint8 == v2.u.v_uint8) ? 0 : ((v1.u.v_uint8 < v2.u.v_uint8) ? -1 : 1);
+        case NNE_TYPE_INT16:
+            return (v1.u.v_int16 == v2.u.v_int16) ? 0 : ((v1.u.v_int16 < v2.u.v_int16) ? -1 : 1);
+        case NNE_TYPE_UINT16:
+            return (v1.u.v_uint16 == v2.u.v_uint16) ? 0 : ((v1.u.v_uint16 < v2.u.v_uint16) ? -1 : 1);
+        case NNE_TYPE_INT32:
+            return (v1.u.v_int32 == v2.u.v_int32) ? 0 : ((v1.u.v_int32 < v2.u.v_int32) ? -1 : 1);
+        case NNE_TYPE_UINT32:
+            return (v1.u.v_uint32 == v2.u.v_uint32) ? 0 : ((v1.u.v_uint32 < v2.u.v_uint32) ? -1 : 1);
+        case NNE_TYPE_STRING:
+            return strcmp(v1.u.v_str, v2.u.v_str);
```

Comment:
You might want to use strncmp here (though I have probably used strcmp lots of time myself :)).